### DISCUSSION TITLE
Forbid Extra Query Parameters

### DISCUSSION THREAD

**Discussion by stevenj** [fastapi/fastapi#9016] [Questions] [answered]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

Similar to the option on a model: 

```python
class MyModel(BaseModel):
    device_id : Optional[str] = Field(None,alias="id")

    class Config:
        """Model Configuration."""

        extra = Extra.forbid
```
Which causes a 422 to be returned if a post body has extra unknown values.  Is it possible to return an error if a query parameter is received on an endpoint which is not defined.

We had an endpoint defined like so:
```python
@endpoint.get("/list")
async def device_list( device_id: Optional[str] = Query(None,alias="id")):
    """Lists a device."""
    return List.device(device_id)
```
Which works fine when someone calls it with the query parameters specified or omited.  But we had a client call it with the query parameter "device_id" not "id"  and they couldn't understand why they kept getting a default response and not the device they were requesting.  The API looked like it was working.  And it looked like the parameters were correct.  But they weren't.

So, is it possible to raise a 422 (or something) if an unknown query parameter is sent in a request to prevent issues like this?

**Comment by aalifadv** [comment]
Would something like this would be what you are looking for?

```python

from pydantic import BaseModel, Field, Extra
from fastapi import FastAPI, Depends


class MyModel(BaseModel):
    device_id: str = Field(alias="id")

    class Config:
        """Model Configuration."""

        extra = Extra.forbid


app = FastAPI()


@app.get("/list")
async def device_list( req_model: MyModel = Depends()):
    """Lists a device."""
    return req_model.device_id
```

```python
In [11]: r = requests.get('http://localhost:8081/list', params={})
In [12]: r.status_code
Out[12]: 422
In [13]: r.json()
Out[13]: 
{'detail': [{'loc': ['query', 'id'],
   'msg': 'field required',
   'type': 'value_error.missing'}]}
In [14]: r = requests.get('http://localhost:8081/list', params={'id':'someid'})
In [15]: r.status_code
Out[15]: 200
In [16]: r.json()
Out[16]: 'someid'
In [17]: r = requests.get('http://localhost:8081/list', params={'notknown':'someid'})
In [18]: r.status_code
Out[18]: 422
```

**Comment by Cleptomania** [comment]
The above works, however I notice you have it as an optional parameter. This way would force it to be a required parameter, rather than checking that they've passed the correct naming to an optional parameter. If you need this to be an optional parameter but need to add a check for people frequently using the wrong naming you can do the below:

You could do this by accessing the Starlette request and examine the parameters dict directly to get a finer level of control over it:

```python
from typing import List, Optional

from fastapi import FastAPI, HTTPException, Query, Request

app = FastAPI()

@app.get("/list")
async def create_item(
    request: Request, device_id: Optional[str] = Query(None, alias="id"), 
):
    params = request.query_params
    if "device_id" in params and "id" not in params:
        raise HTTPException(status_code=422, detail="Please use the 'id' parameter instead of 'device_id'") 

    return device_id
```

The `requests.query_params` is just a python dict containing all of the parameters so you can do just about any validation against that you'd like.

**Comment by stevenj** [comment]
@Cleptomania I am aware of the low level request.  We use it a lot for various things.  I was kind of hoping for something more automatic than manually parsing every endpoint.  

I am now wondering if there would be a way to do it with a dependency that can examine the method, look at the request and compare the Query parameters that are in the request with what the method defines, and automatically reject it if there are any extras.

**Comment by stevenj** [comment]
Answering my own question.  No, it doesn't seem possible without manually processing the request as suggested by @Cleptomania

**Comment by stevenj** [comment]
The only way i could get this to work is i wrote a strictquery dependency function that gets initialized with the names of the query parameters the function can use.  It then throws a validation error like this:

```python
        errors = []

        for query in request.query_params.keys():
            if query not in allowed_params:
                errors.append(ErrorWrapper(ExtraError(), loc=("query", query)))

        if len(errors) > 0:
            raise RequestValidationError(errors=errors, body=request.body)
```            
Its not very DRY because i need to provide the allowed_params list to the dependency which is redundant with the query parameters defined by the endpoint, but i couldn't find a way to collect them automatically.

So, even with that small maintenance issue, it works quite well.

**Comment by Mascarpone** [comment]
You can extend the APIRouter class to get the allowed query parameters.
Thus, with @stevenj code, we get something like this:

```python
from typing import Any, Callable, Optional, Sequence

from fastapi import APIRouter as FastAPIRouter
from fastapi import Depends, params
from fastapi.dependencies.utils import get_dependant
from fastapi.exceptions import RequestValidationError
from fastapi.routing import Request
from pydantic.error_wrappers import ErrorWrapper
from pydantic.errors import ExtraError


async def verify_no_extra_query_param(allowed_params: Sequence[str], request: Request):
    errors = []

    for query in request.query_params.keys():
        if query not in allowed_params:
            errors.append(ErrorWrapper(ExtraError(), loc=("query", query)))

    if len(errors) > 0:
        raise RequestValidationError(errors=errors, body=request.body)


class NoExtraQueryParamChecker:
    def __init__(self, allowed_params: Sequence[str]):
        self.allowed_params = allowed_params

    async def __call__(self, request: Request) -> None:
        await verify_no_extra_query_param(self.allowed_params, request)


class APIRouter(FastAPIRouter):
    def add_api_route(
        self,
        path: str,
        endpoint: Callable[..., Any],
        *,
        dependencies: Optional[Sequence[params.Depends]] = None,
        **kwargs,
    ) -> None:
        """ forbid extra query parameters """
        dependant = get_dependant(path=path, call=endpoint)
        checker = NoExtraQueryParamChecker(dependant.query_params)
        extended_dependencies = [Depends(checker)]
        if dependencies:
            extended_dependencies += [d for d in dependencies]
        super().add_api_route(
            path,
            endpoint,
            dependencies=extended_dependencies,
            **kwargs,
        )

```

**Comment by stevenj** [comment]
@Mascarpone  Thats awesome.

**Comment by j-carson** [comment]
I'd like to request changing this issue from a "question" to a "feature request" or a "documentation bug" -- 

In my opinion, having to hunt down the above code snippet in the github issues doesn't play nice with FastAPI's "we check all that arg stuff for you" philosophy. It should be easier to specify what happens to extraneous query parameters without having to do all the above extra work. If it's not going to be added to the code as a feature, this behavior should be explained somewhere in the documentation.

**Comment by agustinkrebs** [comment]
Hi @Mascarpone! I was trying out your solution (extending the APIRouter class) and find a potential error here:
```python
for query in request.query_params.keys():
    if query not in allowed_params: # ERROR
```
In my case it didn't work because `query` is a string, but `allowed_params` is a list of `ModelField`s. Also, those `ModelField`s could be using aliases, so we have to keep that in mind. I fixed it with:
```python
allowed_params_aliases = {param.alias for param in allowed_params}
for query in request.query_params.keys():
    if query not in allowed_params_aliases:
```
P.S: acessing `param.alias` works even when the user doesn't define a value for it explicitily (it just uses the default value).

**Comment by YannickLeRoux** [comment]
Trying to make the code snippet work but no success so far. 
Seems it can't parse the query params for any endpoint as `request.query_params` is always an empty array.
It results in an error being raised each time I use a query param in my request.

How do have declared your params? I did it this way:

```
class QueryParametersRange(BaseModel):
    """Parameters used to query a range of records within a time window."""

    from_: datetime = Query(
        description="Query records starting from given timestamp. "
        "From must be smaller than to.",
        alias="from",
    )

    to_: datetime = Query(
        description="Query records stopping to given timestamp. "
        "From must be smaller than to.",
        alias="to",
    )
    

___________________________________________________


@router.get(
    "/{acn}/{acc}/range",
    response_model=list[Scenario],
    status_code=status.HTTP_200_OK,
)
def get_scenarios_range(
    *,
    acn: str,
    acc: str,
    params: QueryParametersRange = Depends(),
):
...


___________________________________________________

2 validation errors for Request
query -> from
  extra fields not permitted (type=value_error.extra)
query -> to
  extra fields not permitted (type=value_error.extra)
```

**Comment by JarroVGIT** [comment]
Maybe this will help? Nowadays the `APIRoute` object is included in the scope so you can use that to see what parameters are allowed:

```python
from fastapi import Depends, FastAPI, Request, Query, HTTPException
from fastapi.dependencies.models import Dependant

app = FastAPI()


def strict_query_params(request: Request):
    app: FastAPI = request.app
    requested_path = request.scope["path"]
    dependant: Dependant = request.scope["route"].dependant
    allowed_params = [modelfield.alias for modelfield in dependant.query_params]
    for requested_param, param_value in request.query_params.items():
        if requested_param not in allowed_params:
            raise HTTPException(
                400, f"Not allowed param: {requested_param}: {param_value}"
            )
    return


@app.get("/")
async def root(param1: str = Query(), _=Depends(strict_query_params)):
    return {"hello": "world"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```
The above example will run as-is. Examples using `curl`:
```bash
(venv) jarro@MacBook-Pro-van-Jarro fastapi-github-issues % curl "http://localhost:8000/?param1=text"
{"hello":"world"}%                                                                                                     
(venv) jarro@MacBook-Pro-van-Jarro fastapi-github-issues % curl "http://localhost:8000/?param1=text&id=1"
{"detail":"Not allowed param: id: 1"}%     
```

**Comment by YannickLeRoux** [comment]
Thanks! I like this approach and it works the way you do it (declaring the query params in the function parameters).
But using a pydantic model doesnt work, probably because of my lack of understanding of how it is instantiated by the request. It is useful for me to validate the query params at the model level.
If you try something closer to what I am trying to do, you'll see that `dependant.query_params` is an empty list (but it still parses the query params correctly if removing `strict_query_params`)

```
from fastapi import Depends, FastAPI, Request, Query, HTTPException
from pydantic import BaseModel
from fastapi.dependencies.models import Dependant

app = FastAPI()


def strict_query_params(request: Request):
    app: FastAPI = request.app
    requested_path = request.scope["path"]
    dependant: Dependant = request.scope["route"].dependant
    allowed_params = [modelfield.alias for modelfield in dependant.query_params]
    for requested_param, param_value in request.query_params.items():
        if requested_param not in allowed_params:
            raise HTTPException(
                400, f"Not allowed param: {requested_param}: {param_value}"
            )
    return

class MyParams(BaseModel):
    param1: str | None = Query(None)
    param2: str | None = Query(None)


@app.get("/")
def root(*,params: MyParams =Depends(strict_query_params)):
    if params.param1:
        return {"param1": params.param1}
    return {"hello": "world"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

**Comment by JarroVGIT** [comment, marked as answer]
Fair point, it was just a quick fix. I think this would work for dependencies in models:

```python
from fastapi import Depends, FastAPI, HTTPException, Query, Request
from fastapi.dependencies.models import Dependant
from pydantic import BaseModel

app = FastAPI()


def strict_query_params(request: Request):
    app: FastAPI = request.app
    requested_path = request.scope["path"]
    dependant: Dependant = request.scope["route"].dependant
    allowed_params = [
        param.alias
        for dependency in dependant.dependencies
        for param in dependency.query_params
    ]
    allowed_params += [modelfield.alias for modelfield in dependant.query_params]
    for requested_param, param_value in request.query_params.items():
        if requested_param not in allowed_params:
            raise HTTPException(
                400, f"Not allowed param: {requested_param}: {param_value}"
            )
    return


class MyParams(BaseModel):
    param1: str | None = Query(None)
    param2: str | None = Query(None)


@app.get("/")
def root(*, params: MyParams = Depends(), _=Depends(strict_query_params)):
    if params.param1:
        return {"param1": params.param1}
    return {"hello": "world"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

This should be further refined to enable nested models though, but it's doable. EDIT: Come to think of it; you can't have nested models that are Query parameters.

**Comment by YannickLeRoux** [comment]
It works, thanks a lot for your help!

**Comment by codingjerk** [comment]
Why can't we get this in FastAPI itself?

**Comment by picarro-vmittal** [comment]
We too had similar requirement in our project to restrict any unknown query params in API request. I came up with below work around to raise error in case of unknown query params passed:
```
class QueryParams(BaseModel):
    size: int = DEFAULT_SIZE
    page: int = DEFAULT_PAGE

    class Config:
        extra = Extra.forbid

    def __init__(self, request: Request = None, **kwargs):
        if isinstance(request, Request):
            q_params = dict(request.query_params)
            kwargs.update(q_params)
        super().__init__(**kwargs)

@app.get('/test')
def test(params: QueryParams = Depends()):
     pass
```
This way, it makes sure to pass all query params(known/unknown) to kwargs dict and eventually raises exception due to Extra.forbid config which was otherwise not happening if we were using FastAPI Depends() in api route function

So far, it turns out to be working successfully for our use case. I may not have tested for all corner cases but open for suggestions

**Comment by hajeksc1** [comment]
Looks like forbidding extra query parameters can be done now in a more builtin way in FastAPI (since v0.115.0). The docs say you do it by adding `model_config = {"extra": "forbid"}` to the Pydantic model class definition: https://fastapi.tiangolo.com/tutorial/query-param-models/
